### PR TITLE
Add list-rings

### LINKED ISSUE #521: Add support for shift-click transfer of stacks
In Minecraft, you can shift-click any stack of items in one container and it automatically moves it to another. In a crafting box, this will craft as many of the item that the current materials placed in the 9x9 will make and place it in your inventory. This would be a nifty feature to add to Minetest. Please forgive me if there is already a way to do this in Minetest, I played around with the controls but couldn't find one

### PR DESCRIPTION
**Pull request by est31**
Adds list-rings, a method to implement item sending between inventories via shift-click.
Nice insider feature: a ring consisting of a single inventory list serves as nice clean-up method.
Also adds them to minimal game.
Fixes #521

### CODE REVIEW

**Comment by est31** [comment]
You can notice that shift-clicking too fast "replaces" items instead of transferring them. This is because of lag, and the client trying to send the items to some position that will be occupied by the new items. One solution for this would be a change in protocol, with this as fallback. Another one would involve changing prediction code.

File: doc/lua_api.txt [hunk 1]
```diff
@@ -1340,6 +1340,12 @@ examples.
 #### `list[<inventory location>;<list name>;<X>,<Y>;<W>,<H>;<starting item index>]`
 * Show an inventory list
 
+#### `listring[<inventory location>;<list name>]`
+* Allows to create a ring of inventory lists
+* Shift-clicking on items in one element of the ring
+* will send them to the next inventory list inside the ring
+* The first occurence counts
```
Comment (rubenwardy) [inline doc/lua_api.txt]: the first occur**r**ence of **this tag** overrides all others.

File: games/minimal/mods/default/init.lua [hunk 2]
```diff
@@ -1158,7 +1158,9 @@ minetest.register_node("default:chest", {
 		meta:set_string("formspec",
 				"size[8,9]"..
 				"list[current_name;main;0,0;8,4;]"..
-				"list[current_player;main;0,5;8,4;]")
+				"list[current_player;main;0,5;8,4;]" ..
+				"listring[current_name;main]" ..
+				"listring[current_player;main]")
```
Comment (rubenwardy) [inline games/minimal/mods/default/init.lua]: I'm confused, see above line note.

Comment (est31) [inline games/minimal/mods/default/init.lua]: You basically create a ring where `current_name;main` is followed by `current_player;main` which is again followed by `current_name;main`, like in a ring.

In this example, it means shift-clicking from a chest puts item into the inventory, and vice versa.

You can have more complex examples too, see furnace, where the output slot maps into the inventory, but the inventory maps into the input slot, which itself again maps into the inventory. This is what I've meant with "First occurence counts".

File: doc/lua_api.txt (hunk 1 above)
Comment (est31) [inline doc/lua_api.txt]: it doesnt exactly override, some people might intentionally specify an inventory list multiple times.

**Comment by nerzhul** [comment]
If i understand, you will permit to do shift click on a stack to send it to a chest for example, instead of drag & drop ?

**Comment by nerzhul** [comment]
I don't test the patch but looking at whole code seems good to me, except the multiple pointers which are untested, i comment it to show you

File: src/guiFormSpecMenu.cpp [hunk 3]
```diff
@@ -3308,8 +3349,49 @@ bool GUIFormSpecMenu::OnEvent(const SEvent& event)
...
+					break;
+			}
+			if (i < mis) { // if found
+				u32 to_inv_ind = (i + 1) % mis;
+				const ListRingSpec &to_inv_sp = m_inventory_rings[to_inv_ind];
+				InventoryList *list_from = inv_s->getList(s.listname);
+				Inventory *inv_to = m_invmgr->getInventory(to_inv_sp.inventoryloc);
+				InventoryList *list_to = inv_to->getList(to_inv_sp.listname);
+				ItemStack stack_from = list_from->getItem(s.i);
+				assert(shift_move_amount <= stack_from.count);
```
Comment (nerzhul) [inline src/guiFormSpecMenu.cpp]: I think we should test the pointers before this line.

**Comment by est31** [comment]
Yes it will permit that.

**Pull request merged**